### PR TITLE
Make RegexCompiler-generated code more JIT friendly

### PR DESCRIPTION
**Pull request by stephentoub**
- Prefer using stloc/ldloc over using dup, especially across branches.
- Reduce the depth of the evaluation stack in a few key places.
- Use call instead of callvirt where applicable.
- Use the shorter form of some instructions where applicable.

On a customer-provided workload featuring several hundred complex regexes, this improves the time it takes to JIT the RegexCompiler-generated code and the time it takes to perform a bunch of matches by 15-20%.

cc: @AndyAyersMS, @eerhardt, @pgovind

### CODE REVIEW

**Review by AndyAyersMS — COMMENTED** [review]
There are still a couple places that call `Dup`, though from what I can tell they won't result in multiple instances of `dup` in the IL stream, so there may not be much benefit from reworking those.

**Comment by stephentoub** [comment]
> There are still a couple places that call Dup,

Only in FindFirstChar rather than Go, right?  The latter can get huge; the former is generally much more constrained.

**Comment by stephentoub** [comment]
Ah, there are still one or two in the newer Go implementation I added.  I can look at removing those, too.

**Comment by stephentoub** [comment]
Thanks for the push, Andy.  I decided to just go and delete the rest of the dups :)

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs [hunk 1]
```diff
@@ -4904,19 +4919,19 @@ private void GenerateOneCode()
...
-                            Ldc(c);
-                            Blt(l4);
-                            Pop();
-                            Ldc(c);
-                            MarkLabel(l4);
-                        }
                         using (RentedLocalBuilder cLocal = RentInt32Local())
                         {
                             Stloc(cLocal);
+                            if (c != int.MaxValue)
```
Comment (danmosemsft) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:4925]: Aside - the commented blocks of equivalent code seem to be pretty out of sync in some cases. Eg., `int.MaxValue` doesn't appear in the block above. I wonder whether it is worth (separately) updating those as far as possible. It is generally not possible to infer what is going on by comparing the interpreter.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:4925]: Yeah, I've added and cleaned up existing comments as I go, but there's still a non-trivial amount of code in the library that could stand to have better / revised comments.

**Review by eerhardt — APPROVED** [review]
Looks good to me.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs [hunk 2]
```diff
@@ -671,10 +674,13 @@ private void PopTrack()
             Ldloc(_runtrackLocal!);
             Ldloc(_runtrackposLocal!);
             LdelemI4();
+            using RentedLocalBuilder tmp = RentInt32Local();
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:677]: > Are these 2 of the "Reduce the depth of the evaluation stack in a few key places." places?

Yup.  On the repro I have, these two have a measurable impact on the JIT's throughput.

**Pull request merged**